src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,449 @@
+              std::vector<Predicate> dlir_rhs = {};
+              for (auto ast_rhs : conditional_assertion.rhs()) {
+                // extra rule are only generated for facts on the LHS,
+                // so the rules that would be generated from this RHS fact are
+                // not used.
+                auto [dlir_translation, not_used] =
+                    BaseFactToDLIR(speaker, ast_rhs);
+                dlir_rhs.push_back(
+                    PushPrin(std::string("says_"), speaker, dlir_translation));
+              }
+              auto [dlir_lhs, gen_rules] =
+                  FactToDLIR(speaker, conditional_assertion.lhs());
+              auto dlir_lhs_prime =
+                  PushPrin(std::string("says_"), speaker, dlir_lhs);
+              auto dlir_assertion =
+                  DLIRAssertion(DLIRCondAssertion(dlir_lhs_prime, dlir_rhs));
+              gen_rules.push_back(dlir_assertion);
+              return gen_rules;
+            }},
+        assertion.GetValue());
+  }
+
+  std::vector<DLIRAssertion> SaysAssertionToDLIR(
+      const SaysAssertion& says_assertion) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const Assertion& assertion : says_assertion.assertions()) {
+      std::vector<DLIRAssertion> single_translation =
+          SingleSaysAssertionToDLIR(says_assertion.principal(), assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
+    }
+    return ret;
+  }
+
+  std::vector<DLIRAssertion> SaysAssertionsToDLIR(
+      const std::vector<SaysAssertion>& says_assertions) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const SaysAssertion& says_assertion : says_assertions) {
+      auto single_translation = SaysAssertionToDLIR(says_assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
+    }
+    return ret;
+  }
+
+// Queries are like predicates with arity 0. Souffle does not have predicates
+// with arity 0, so we represent them as having one argument which is a
+// constant.
+#define DUMMY_PREDICATE Predicate("grounded_dummy", {"dummy_var"}, kPositive)
+
+  DLIRAssertion QueryToDLIR(const Query& query) {
+    // The assertions that are normally generated during the translation
+    // from facts to dlir facts are not used for queries.
+    auto [main_pred, not_used] = FactToDLIR(query.principal(), query.fact());
+    main_pred = PushPrin(std::string("says_"), query.principal(), main_pred);
+    auto lhs = Predicate(query.name(), {"dummy_var"}, kPositive);
+    return DLIRAssertion(DLIRCondAssertion(lhs, {main_pred, DUMMY_PREDICATE}));
+  }
+
+  std::vector<DLIRAssertion> QueriesToDLIR(const std::vector<Query>& queries) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const Query& query : queries) {
+      ret.push_back(QueryToDLIR(query));
+    }
+    return ret;
+  }
```

Comment:
Might be nicer to just do the following:

1. Have `queries` be of value type `std::vector<Query>`, introducing an elidable copy at calls to the function.
2. Have `ret` be populated via an `insert` at the end, like so:
```C++
insert(ret.end(), std::make_move_iterator(queries.begin()), std::make_move_iterator(queries.end()):
```

(I did not look up the `insert` function, syntax might be slightly wrong).

Reply:
Can't make queries be of value type because program is a const reference since it should not be moved.